gapic/src/main/com/google/gapid/models/Memory.java

```diff
@@ -641,38 +644,21 @@ private boolean mayHaveChildren() {
     /**
      * Utility method. Simplify trees, especially for vulkan structs.
      * 1. Remove redundant layers for all the trees.
-     * 2. Identify some trees to be the main trees. (Assume the main trees to be those with type
-     *    TypeInfo.StructType, they usually contain key info like VkPresentInfoKHR, VkSubmitInfo...)
-     * 3. Combine trees together by appending some smaller trees to the main trees, if they are
+     * 2. Combine trees together by appending some smaller trees to the main trees, if they are
      *    related through a pointer field.
      */
     public static List<StructNode> simplifyTrees(List<StructNode> trees) {
-      List<StructNode> simplified = new ArrayList<StructNode>();
-
+      // Remove redundant layers.
       Map<Long, StructNode> nodes = new HashMap<Long, StructNode>();
       for (StructNode tree : trees) {
         nodes.put(tree.getRootAddress(), removeExtraLayers(tree));
       }
-
-      // Find the main trees.
-      for (Iterator<Map.Entry<Long, StructNode>> it = nodes.entrySet().iterator(); it.hasNext(); ) {
-        Map.Entry<Long, StructNode> entry = it.next();
-        if (containsStructType(entry.getValue())) {
-          simplified.add(entry.getValue());
-          it.remove();
-        }
-      }
-
-      // Append other trees to the main trees if possible.
-      for (StructNode mainTree : simplified) {
-        appendPointedNodes(mainTree, nodes);
-      }
-
-      // Add the remaining unappended nodes to the returning result.
+      // Append pointed nodes to corresponding pointer field if possible.
+      Set<StructNode> appended = Sets.newHashSet();
       for (StructNode node : nodes.values()) {
-        simplified.add(node);
+        appendPointedNodes(node, nodes, appended);
       }
-      return simplified;
+      return nodes.values().stream().filter(n -> !appended.contains(n)).collect(Collectors.toList());
```

Comment:
I think this is a pretty big hammer... but OK

Reply:
The only real alternative would be to make a copy first and have `appendPointedNodes` remove from the copy.